builtin/logical/database/backend.go

```diff
@@ -228,6 +272,156 @@ func (b *databaseBackend) GetConnection(ctx context.Context, s logical.Storage,
+// the WAL. If the matching Role does not exist, the Role's LastVaultRotation is
+// newer than the WAL, or the Role does not have a static account, delete the
+// WAL.
+func (b *databaseBackend) loadStaticWALs(ctx context.Context, conf *logical.BackendConfig) error {
+	keys, err := framework.ListWAL(ctx, conf.StorageView)
+	if err != nil {
+		return err
+	}
+	if len(keys) == 0 {
+		b.Logger().Debug("no WAL entries found")
+		return nil
+	}
+
+	// loop through WAL keys and process any rotation ones
+	var merr error
+	for _, walID := range keys {
+		// allow cancellation from context
+		select {
+		case <-ctx.Done():
+			b.Logger().Info("loading WAL entries cancelled")
+			return merr
+		default:
+		}
+
+		walEntry := b.findStaticWAL(ctx, conf.StorageView, walID)
+		if walEntry == nil {
+			continue
+		}
+
+		// load matching role and verify
+		role, err := b.StaticRole(ctx, conf.StorageView, walEntry.RoleName)
+		if err != nil {
+			b.Logger().Warn("error loading role", err)
+			merr = multierror.Append(merr, err)
+			continue
+		}
+
+		if role == nil || role.StaticAccount == nil {
+			b.Logger().Warn("role or static account not found")
+			if err = framework.DeleteWAL(ctx, conf.StorageView, walID); err != nil {
+				b.Logger().Warn("error deleting WAL for role with no static account", err.Error())
+				merr = multierror.Append(merr, err)
+			}
+			continue
+		}
+
+		if role.StaticAccount.LastVaultRotation.After(walEntry.LastVaultRotation) {
+			// role password has been rotated since the WAL was created, so let's
+			// delete this
+			if err = framework.DeleteWAL(ctx, conf.StorageView, walID); err != nil {
+				b.Logger().Warn("error deleting WAL for role with newer rotation date", err.Error())
+				merr = multierror.Append(merr, err)
+			}
+			continue
+		}
+
+		// setStaticAccount which will attempt to set the password and
+		// delete the WAL if successful
+		resp, err := b.setStaticAccount(ctx, conf.StorageView, &setStaticAccountInput{
+			RoleName: walEntry.RoleName,
+			Role:     role,
+			WALID:    walID,
+			Password: walEntry.NewPassword,
+		})
+		if err != nil {
+			// if response contains a WALID, create an item to push to the queue with
+			// a backoff time and include the WAL ID
+			merr = multierror.Append(merr, err)
+			if resp.WALID != "" {
+				// Add their rotation to the queue
+				if err := b.pushItem(&queue.Item{
+					Key:      walEntry.RoleName,
+					Value:    walID,
+					Priority: walEntry.LastVaultRotation.Add(time.Second * 60).Unix(),
+				}); err != nil {
+					b.Logger().Warn("error pushing item on to queue after failed WAL restore", err)
+					merr = multierror.Append(merr, err)
+				}
+			}
+		}
+	} // end range keys
+	return merr
+}
+
+// invalidateQueue cancels any background queue loading and destroys the queue.
+func (b *databaseBackend) invalidateQueue() {
+	b.Lock()
+	defer b.Unlock()
+
+	// cancelQueue
+	if b.cancelQueue != nil {
+		b.cancelQueue()
+	}
+	b.credRotationQueue = nil
```

Comment:
Supposedly after we empty cancel and set to `nil` we need to repopulate the queue with data? Although...since only the active node on the primary cluster will have the queue populated i'm not sure we need an invalidation function at all

Reply:
Should clarify we do still need a way to stop the queue on `Clean`, but in that case we do not need to think about repopulating the queue. So the logic here is likely correct, we just can remove this call from the invalidation handler.

Reply:
I did remove the call to `invalidateQueue` from the clean methods. I also added new methods to the backend, `popItem` and `popItemByKey` to guard against a nil `credRotationQueue` being hit in `rotateCredentials` which is running in a separate go-routine